### PR TITLE
Support for polygons with holes in QgsRubberBand

### LINKED ISSUE #34399: Polygonal rubber bands cannot have holes
**Describe the bug**
When creating a polygon rubber band, only the outer polygon ring is taken into account. As a result one cannot create a rubber band with the geometry of a polygon with holes.

**How to Reproduce**
- Create a polygon and use the fill ring tool to punch a hole in the middle.
- Select both geometries and click merge attributes.
- When clicking on the attributes row a rubber band is created to highlight the selected feature. The rubber band for the feature with the hole is displayed with it's hole filled.

**QGIS and OS versions**
Master on Debian

**Additional context**
Only the first ring is used, similarly for MultiPolygons
https://github.com/qgis/QGIS/blob/2aba793946537a2555a463d923c9e7913ac5d138/src/gui/qgsrubberband.cpp#L337-L345

As a result passing a polygon with a hole to `QgsRubberBand::setToGeometry()`  then calling  `QgsRubberBand::asGeometry()` does not return the same polygon.

### PR DESCRIPTION
**Pull request by uclaros**
## Description
Modified `QgsRubberBand` to correctly display polygons with holes (inner rings) that are created by  `QgsRubberBand::addGeometry()` / `QgsRubberBand::setToGeometry()`.
This affects the Vertex tool, Merge tool and anywhere else a rubberband is created from an existing geometry.
While at it, I changed from using `QList` to `QVector` since it's used in `Qgs*****XY` and from using stl method names to Qt ones (`.append()` instead of `.push_back()` etc since they are used more often elsewhere in QGIS.

I am not sure about `QgsRubberBand::partSize()` though. Should it continue to return the vertex count of the outer ring only, or should it return the total vertices in all rings (in case of polygons with holes)? It's only used on the unit tests as far as I can tell.
https://github.com/uclaros/QGIS/blob/dc09b9d752ece40e2c744c7576f69fc196d666d3/src/gui/qgsrubberband.cpp#L635-L639

I can add a couple of tests if all is OK.

Fixes #34399

### CODE REVIEW

**Comment by m-kuhn** [comment]
The way I understood QgsRubberBand is that it's a utility to manage rings.
Moving this further the way you propose opens up a number of edge cases e.g. what should be done if a rubberband is applied to a "hole" ring inside a polygon.

For visualisation purpose, the class "QgsHighlight" is available.

I would prefer to improve the docs (and API) to make the usage of QgsRubberBand as ring management tool clearer.
Unless I have missed some critical missing parts of QgsRubberBand and QgsHighlight.

**Comment by uclaros** [comment]
`QgsRubberBand` is about visualizing live user input while adding, moving or removing nodes. It's used when digitizing points, multipoints, lines, multilines, polygons and multipolygons to give visual feedback about what is being digitized. While digitizing new geometries, we don't need to care about inner rings, since they can only be created on a second step with the *add ring* tool, that's why they were initially ignored by `QgsRubberBand`. *Vertex tool*, on the other hand, handles complete geometries that may already have inner rings while still require to give that visual feedback to the user and this is where `QgsRubberBand` was lacking. Also when digitizing inner rings, the *add ring* tool can now be adapted to give proper feedback and a live preview of the hole being punched.

The case of the *merge selected features* tool is different though and you are right there. The geometries when merging are fixed, so I totally agree that `QgsHighlight` should have been used instead of a `QgsRubberBand`.

As for the edge cases you mention, it works the other way round. You don't put a `QgsRubberBand` on a ring, but a ring to a `QgsRubberBand`. A single inner ring is not a hole any more since it's single, it's an outer ring now! So putting an inner ring to a `QgsRubberBand` it will be displayed as an outer ring, and that's expected behavior. Maybe we could add a method to add inner rings to an existing `QgsRubberBand` though to be used with the afforementioned *add ring* tool adaptation.

**Comment by m-kuhn** [comment]
So this is mainly about UX improvements for the vertex tool?
How do we deal with multi polygons?

**Comment by uclaros** [comment]
> So this is mainly about UX improvements for the vertex tool?

The vertex tool is just how I discovered #34399 because of heavy daily usage.
IMHO `QgsRubberBand` is broken right now. It has an `addGeometry()` method to feed it with geometries and a `asGeometry()` method that does not return the geometry it was fed with and also does not draw it correctly if it has inner rings!
The merge tool will also benefit from fixing it until it's modded to use `QgsHighlight` instead of `QgsRubberBand`.

>  How do we deal with multi polygons?

They were already handled by storing the rubber band points in a `QList< QList<QgsPointXY> >`. So adding a nesting level enables it to handle inner rings too.

**Comment by uclaros** [comment]
Any comments / questions regarding this are welcome; I'm keeping it alive.

**Comment by cxcandid** [comment]
@uclaros: +1 ... this is highly awaited, at least by me ... thanks for your proposal!

Don't need my workaround anymore :-)

```
    def showRubberBand(self,wkt,srcSRS,destSRS,move=False):
        # remove welcome page
        if self.iface.mapCanvas().isHidden():
            self.iface.newProject()

        xform = QgsCoordinateTransform(srcSRS,destSRS,QgsProject.instance())
        regex = re.compile(r"^POLYGON\s*", re.IGNORECASE)

        wkt = regex.sub("",wkt)
        ptlist = eval(wkt.replace(", ",",").replace("((","[(").replace("))",")]").replace("(","[(").replace(")",")]").replace("],[","]#[").replace(",","),(").replace(" ",",").replace("]#[","],["))

        points = [[]]
        boundary = ptlist.pop(0)
        pt0 = xform.transform(QgsPointXY(boundary[0][0],boundary[0][1]))

        for pt in boundary:
          points[0].append(xform.transform(QgsPointXY(pt[0],pt[1])))

        points[0].append(QgsPointXY(pt0[0],pt0[1]))
        geom = QgsGeometry.fromPolygonXY(points)

        self.rb.reset()
        self.rb.setWidth(0)
        self.rb.setStrokeColor(QtGui.QColor.fromRgb(0,0,0,0))
        self.rb.setToGeometry(geom,None)
        
        for ring in ptlist:
            for pt in ring:
                self.rb.addPoint(xform.transform(QgsPointXY(pt[0],pt[1])))
            self.rb.addPoint(pt0)
        
        if move:
            box = geom.boundingBox()
            self.iface.mapCanvas().setExtent(box)
            self.iface.mapCanvas().zoomOut()
            self.iface.mapCanvas().refresh()
```

**Comment by uclaros** [comment]
Some love for rubberband donuts?

**Review by m-kuhn — COMMENTED** [review]
Sorry for the delay. Given that we already have support for multi geometries, I think the donuts are ok to support too.

I'm happy with merging this if no one has any objections.

File: src/gui/qgsrubberband.cpp [hunk 1]
```diff
@@ -173,14 +178,14 @@ void QgsRubberBand::removePoint( int index, bool doUpdate/* = true*/, int geomet
   }
 
 
-  if ( !mPoints[geometryIndex].isEmpty() )
+  if ( !mPoints.at( geometryIndex ).at( ringIndex ).isEmpty() )
```
[the remark this answers is not in the document]
Comment (uclaros) [reply src/gui/qgsrubberband.cpp]: Added index range checks for all `.at()` and `[]` accessors.

File: python/gui/auto_generated/qgsrubberband.sip.in [hunk 2]
```diff
@@ -315,12 +318,13 @@ Returns count of vertices in all lists of mPoint
 :return: The total number of vertices
 %End
 
-    const QgsPointXY *getPoint( int i, int j = 0 ) const;
+    const QgsPointXY *getPoint( int i, int k = 0, int j = 0 ) const;
```
[the remark this answers is not in the document]
Comment (uclaros) [reply python/gui/auto_generated/qgsrubberband.sip.in]: Also reverted the parameter order and a note to be fixed. Should I open a ticket on QGIS 4 or is the `// TODO QGIS 4:` enough?

[the remark this answers is not in the document]
Comment (nyalldawson) [reply python/gui/auto_generated/qgsrubberband.sip.in]: The TODO is fine

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**